### ISSUE TITLE
Memory leak when using actionChannel() with takeLatest()

### ISSUE THREAD

**Issue by cgav** [redux-saga/redux-saga#1598]
#### Description of the bug/issue

I have observed a 'channel overflow' error using `redux-saga@0.16.0` when running the following code snippet. After upgrading to `redux-saga@1.0.0-beta.1` the overflow disappears but the channels are not being closed after `takeLatest()` is (or should have been) cancelling `takeLatestSaga`:

```javascript
const actionChannelSaga = function*(id) {
  console.log(`### Channel saga for ACTION_FOR_TAKE_LATEST ${id}`);
};

const takeLatestSaga = function*(action) {
  const { id } = action.payload;
  const channel = yield actionChannel('ACTION_FOR_CHANNEL');
  
  const originalPut = channel.put;
  channel.put = (...args) => {
    console.log(`channel.put() called for id ${id}`);
    originalPut(...args);
  };

  while (true) {
    const action = yield take(channel);
    yield call(actionChannelSaga, id, action);
    yield take('ACTION_THAT_NEVER_COMES');
    console.log('>>> We should not reach this point');
  }
};

const startSaga = function*() {
  yield takeLatest('ACTION_FOR_TAKE_LATEST', takeLatestSaga);

  for (let i = 0; i < 10; i++) {
    yield put({ type: 'ACTION_FOR_TAKE_LATEST', payload: { id: i } });
    yield call(delay, 10);
    yield put({ type: 'ACTION_FOR_CHANNEL' });
  }
};

sagaMiddleware.run(startSaga);
```

#### Steps to reproduce the bug/issue

1. Run the snippet above (see codesandbox.io link below)
...

#### Example
https://codesandbox.io/s/vn3ljqq9y0

### Actual results

The snippet above produces the following logs:
```
channel.put() called for id 0
### Channel saga for ACTION_FOR_TAKE_LATEST 0
channel.put() called for id 0
channel.put() called for id 1
### Channel saga for ACTION_FOR_TAKE_LATEST 1
channel.put() called for id 0
channel.put() called for id 1
channel.put() called for id 2
### Channel saga for ACTION_FOR_TAKE_LATEST 2
channel.put() called for id 0
channel.put() called for id 1
channel.put() called for id 2
channel.put() called for id 3
### Channel saga for ACTION_FOR_TAKE_LATEST 3
channel.put() called for id 0
channel.put() called for id 1
channel.put() called for id 2
channel.put() called for id 3
channel.put() called for id 4
### Channel saga for ACTION_FOR_TAKE_LATEST 4
...
```

### The Expected results

Running the snippet above, I'd expect the `actionChannel` to be closed after `takeLatestSaga()` is being cancelled and to see those logs instead:
```
channel.put() called for id 0
### Channel saga for ACTION_FOR_TAKE_LATEST 0
channel.put() called for id 1
### Channel saga for ACTION_FOR_TAKE_LATEST 1
channel.put() called for id 2
### Channel saga for ACTION_FOR_TAKE_LATEST 2
channel.put() called for id 3
### Channel saga for ACTION_FOR_TAKE_LATEST 3
channel.put() called for id 4
### Channel saga for ACTION_FOR_TAKE_LATEST 4
...
```

### Environment information
- redux-saga 1.0.0-beta.1
- redux 3.7.2

**Comment by shinima** [comment]
Nice catch. The problem is related to the following code: https://github.com/redux-saga/redux-saga/blob/63dfb94ebfaadf1b59a3e4029a1fa71c574307a4/packages/core/src/internal/proc.js#L638-L652

Since we never dispatch END into env.stdChannel, `env.stdChannel.take(taker, match)` always gets executed even when the actionChannel is closed, so `channel.put` is still called after the channel being closed.

**Comment by Andarist** [comment]
I've looked into this now briefly, but I don't see how END relates to the problem - it's never dispatched in this scenario.

The `chan.close` is also never called automatically because `yield actionChannel()` just connects to `stdChannel`, but it doesn't actually register itself to the task, so when task gets cancelled it has 0 chance to close the channel (it simply is not aware of it).

We could track created `actionChannel`s and close them on cancellation, but we would have to think through if there are any downsides of closing them automatically - like is there any possible scenario when one wouldnt want such behaviour? Remember that created channels can be passed around, but it seems to me that without abusing context feature they can only be passed down the task tree, so it should be quite safe to assume that if the creating task gets cancelled it should close its actionChannels (maybe we could extend the behaviour for all channels later, but it would require introducing a new effect for channel creation).

**Comment by shinima** [comment]
Not sure whether we should close the channels automatically 😕 . Users could close the channels manually in the finally-block if they want to ensure the channels being closed when exiting task.

The current issue I found is that even if we call `channel.close()` explicitly, the memory leak still occurs. See [this codesandbox](https://codesandbox.io/s/9j926mv424?module=%2Fsrc%2Fsagas%2Findex.js)

**Comment by Andarist** [comment]
Yeah, you are right - we should cancel the taker on `stdChannel` in that case. Good thing is that this `put` is effectively a `noop` - https://github.com/redux-saga/redux-saga/blob/9523169a31d4ae90d00a6adc3482697478c2271a/packages/core/src/internal/channel.js#L176-L178 , but ofc it doesn't free up resources and we need to fix that.

> Not sure whether we should close the channels automatically 😕 . Users could close the channels manually in the finally-block if they want to ensure the channels being closed when exiting task.

Can u think of any use case for `actionChannel` that would make auto closing undesirable?

**Comment by shinima** [comment]
I found that auto closing is undesirable when we use `spawn` effects. The detached children may still use a channel from the parent while the parent task is complete.

**Comment by Andarist** [comment]
fixed by #1606 , as mentioned we cannot call close automatically, but at least we've fixed a memory leak thanks to your report ❤️
